WIP: Add db

Flaque opened the pull request:
# ✨ Overview 

This adds a database to metapod. Since this is a pretty important and formal piece of the project, I'm doing this entirely in a PR so it's obvious what's going on.

# 🛠 Tech Choices

All of these are up for debate. 

## Database

Postgres. No weird dev environment, good performance, often used. 

## Migrations

For migrations, I'm fond of [golang-migrate/migrate](https://github.com/golang-migrate/migrate). It's capable of being used in other languages, but also has a go library. It works with tons of dbs and is easy to setup. 

# 📐 ERD / Tables 
<img width="363" alt="screen shot 2018-09-24 at 9 25 04 pm" src="https://user-images.githubusercontent.com/5942769/45992790-7d741680-c040-11e8-9fcf-85437a677f2b.png">

To best follow the spec, we only need a pretty minimal table set at the moment. 

## Group

A Group corresponds to a [Group actor](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-group) in the ActivityStreams vocabulary.

As such, a Group has many Organizations. 

An example Group might be "NPR" because NPR has many Organizations (Fresh Air, Planet Money, All Things Considered, Car Talk, etc) that run podcasts. 

| name | type |
|------|------|
| id   | int  |
| name | string |
| notes | html | 
| createdAt | date |
| updatedAt | date |

## Organizations

An Organization corresponds to an [Organization actor](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-organization) in the ActivitySteams vocabulary.

An Organization has many Podcasts. 

An example Organization might be "Planet Money".

| name | type |
|------|------|
| id   | int  |
| name | string |
| notes | html | 
| groupId | int |
| createdAt | date |
| updatedAt | date |

## Podcast 
> Potential other names: "Show", "Episode"

A Podcast is an Object that has a URL link to an Audio file, an HTML blob for show notes, a thumbnail, a name, and a date.

An example podcast might be "#444: New Jersey Wine". 

| name | type |
|------|------|
| id   | int  |
| name | string |
| notes | html | 
| thumbnailUrl | string |
| audioUrl | string |
| mediaType | string |
| organizationId | int |
| postedAt | date |
| createdAt | date |
| updatedAt | date |

Note that podcasts can often be scheduled to release at particular times, hence why there's a `postedAt` time along side the traditional `createdAt` and `updatedAt`. 

(Not sure if we actually need the media type or if we can get that from a URL).

### Potential considerations
It's common for podcasts to rerelease episodes, we may therefore wish to not make titles unique or provide some way other way to rerelease podcasts. 

# 🤔 Open Questions

## Slugs?

It may be beneficial for us to consider using slugs as a unique identifier for some tables. A slug would be generated from some table attribute, typically the name in most cases.

So for example, a podcast titled "#444: New Jersey Wine" might become "new-jeresy-wine" or maybe "444-new-jersey-wine". Libraries like [gosimple/slug](https://github.com/gosimple/slug) make it easy for us to convert international characters into easy ascii urls:

```go
someText := slug.Make("影師")
fmt.Println(someText) // Will print: "ying-shi"
```

In the case of rereleases, we could just add on a +1 to the slug:

```
444-new-jersey-wine-1
```

Unique slugs often lead to a clean, debuggable, and user-friendly api. For example, which of the following is easier to reason about?

```
/activity/organization/1230412/podcast/00123023
/activity/organization/this-american-life/podcast/new-jeresy-wine
```

Slugs aren't required to make urls like these, but they give us something to index on and encourage good practices.

On the other hand, slugs can also lead to several sources of truth. Without good discipline, it's often the case that different developers will use ids in some places and slugs in others which can lead to a less clean API.

Flaque committed:
Add db

Commit:
Containerized api and db using docker-compose.
Modified Makefile to use docker-compose.
No longer need to build dependencies localy, development will take place within docker-compose.

Commit:
Updated travis CI to just build the docker container (runs tests)

maxchehab commented:
@Flaque I think slugs are a good idea... would we add this as a field for each model?

Flaque commented on .gitignore:
```diff
@@ -21,8 +21,9 @@
 .DS_Store
 *.swp
 
-# binary
```
FIX: We should still keep this in the gitignore because someone somewhere is going to run `go install` and the accidentally commit this. 🤷‍♀️

Flaque commented on .gitignore:
```diff
@@ -21,8 +21,9 @@
-# binary
-metapod
+#docker-composer
+tmp
```
Does docker compose add a tmp file or something?

maxchehab replied:
My bad, fresh actually creates to store their binary and error logs. https://github.com/pilu/fresh#usage

Flaque commented on README.md:
```diff
@@ -6,30 +6,28 @@ An _experimental_ (Read: not-usable or in anyway done) distributed podcasting pl
 
 Getting started:
 
-Install the vendor dependencies using godep. (https://github.com/golang/dep)
```
FIX: Keep this part in, it helps document to less-experienced go folks (like myself 2 days ago 😅) that we're using godep and not govendor.

Flaque commented on README.md:
```diff
@@ -6,30 +6,28 @@ An _experimental_ (Read: not-usable or in anyway done) distributed podcasting pl
 Getting started:
 
-Install the vendor dependencies using godep. (https://github.com/golang/dep)
+We use [fresh](https://github.com/pilu/fresh) for hot code reloading with docker-compose.
```
Doesn't fresh only work with traffic, martini or gocraft/web? We may be able to use that, though at the moment we're not.

maxchehab replied:
Fresh works with any changes.

Flaque replied:
^ Dope.

Flaque commented on README.md:
````diff
@@ -6,30 +6,28 @@ An _experimental_ (Read: not-usable or in anyway done) distributed podcasting pl
+Building docker containers for production:
 
 ```sh
-make test
````
FIX: Keep `make test` separate from making prod. It encourages people to write and run tests often and at the moment. They should be as fast as possible to run.

Flaque commented on db/.gitignore:
```diff
@@ -0,0 +1 @@
+pgdata
```
NIT: Unless there's a strong reason for having another gitignore, I'd prefer to use only one out of convention.

Flaque reviewed (commented):
This is super cool! 🙌 I left comments/change-requests though 😅

Flaque commented:
> I think slugs are a good idea... would we add this as a field for each model?

@maxchehab, yeah basically. Just a tinystring on each one.

Commit:
Created specific testing command.
Updated README.md to display testing command and dep vendor information.
Updated travis.yml to use specific testing command
Combined database gitignore with repositories gitignore.:

Flaque commented:
Also @maxchehab, don't delete the `vendor` folder. I'd prefer to keep it in for this project. At the moment, there's no way for golang to handle someone else removing a library. [See the dep/FAQ for more reasoning](https://github.com/golang/dep/blob/master/docs/FAQ.md#should-i-commit-my-vendor-directory)

That said, I'm willing to change my mind on this. 

Also, I'm fine moving stuff into an `api` package, but why did we move the `toml` and `lock` into there? that's not package specific right?

maxchehab commented:
@Flaque my thought process with removing `vendors/` was that all development would be conducted within docker-compose. This means the only requirements for the developer are extremely low (just have to have docker)... this is also why I put the `toml` and `lock` in the api folder. Everything within the api folder should be sufficient to build a docker image for the api.

Flaque commented:
Like all other things, this PR has been partially migrated to [here](https://github.com/metapods/metapods/pull/1)
